smack-websocket/src/main/java/org/jivesoftware/smack/websocket/WebsocketConnectionAttemptState.java

```diff
@@ -0,0 +1,95 @@
+import org.jivesoftware.smack.websocket.implementations.OkHttpWebsocketImpl;
+import org.jivesoftware.smack.websocket.implementations.WebsocketImplProvider;
+import org.jivesoftware.smack.websocket.implementations.WebsocketInterface;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+
+public final class WebsocketConnectionAttemptState {
+
+    private final ModularXmppClientToServerConnectionInternal connectionInternal;
+    private final XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints discoveredEndpoints;
+
+    final List<RemoteConnectionException<?>> connectionExceptions;
+    EndpointConnectionException connectionException;
+
+    final Iterator<WebsocketRemoteConnectionEndpoint> connectionEndpointIterator;
+    WebsocketRemoteConnectionEndpoint connectionEndpoint;
+    Iterator<? extends InetAddress> inetAddressIterator;
+
+    WebsocketInterface websocket;
+
+    WebsocketConnectionAttemptState(ModularXmppClientToServerConnectionInternal connectionInternal,
+                    XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints,
+                    EstablishingWebsocketConnectionState establishingWebsocketConnectionState) throws IOException {
+        this.connectionInternal = connectionInternal;
+        this.discoveredEndpoints = discoveredWebsocketEndpoints;
+
+        List<WebsocketRemoteConnectionEndpoint> endpoints = discoveredWebsocketEndpoints.result.discoveredRemoteConnectionEndpoints;
+        connectionEndpointIterator = endpoints.iterator();
+        connectionEndpoint = connectionEndpointIterator.next();
+        connectionExceptions = new ArrayList<>(endpoints.size());
+    }
+
+    StateTransitionResult.Failure establishWebsocketConnection() throws InterruptedException {
+
+        try {
+            websocket = WebsocketImplProvider.getWebsocketImpl(OkHttpWebsocketImpl.class, connectionInternal, discoveredEndpoints);
+            for (WebsocketRemoteConnectionEndpoint endpoint : discoveredEndpoints.getResult().discoveredRemoteConnectionEndpoints) {
+                boolean connectionSuccessful = websocket.connect(endpoint);
+                if (connectionSuccessful) {
+                    break;
+                }
+            }
+        } catch (NoSuchMethodException | SecurityException | InstantiationException |
+                IllegalAccessException | IllegalArgumentException | InvocationTargetException exception) {
+            Logger.getAnonymousLogger().log(Level.WARNING, "Couldn't obtain the underlying websocket implementation: ", exception);
+            return new StateTransitionResult.FailureCausedByException<Exception>(exception);
+        } catch (SmackException | XMPPException | WebsocketException exception) {
+            Logger.getAnonymousLogger().log(Level.WARNING, "Couldn't connect to any of the discovered endpoints: ", exception);
```

Comment:
You perform the exact same code in both `catch` blocks. There is no point in doing so, isn't it?

I think what you really want to do is to return if `getWebsocketImpl()` throws, but collect the exceptions thrown by `connect()`, and only return unsuccessfully of all connection endpoints fail.